### DISCUSSION TITLE
Welcome to yowsup Discussions!

### DISCUSSION THREAD

**Discussion by tgalal** [tgalal/yowsup#3027] [General]
## Welcome!

Ask questions you’re wondering about and share ideas. Advertisements or asking for paid help are not allowed.

**Comment by twilliamftv** [comment]
@tgalal Is there a solution to send media?

**Reply by ben221199** [reply]
What errors do you get when trying to send media? If so, you should look if there is any issue that fixes it. If there is none, you can create your own issue.

**Comment by zek** [comment]
Greetings @tgalal, 

- What tools/docs do you recommend for reverse engineering? Do you bypass SSL pinning or any security wall.
- How do you see the future of yowsup?

**Reply by ben221199** [reply]
- When you bypass the SSL pinning and capture the data with a MITM proxy, you will find how the registration flow works and that after that, the WhatsApp client sends `WA` `0x03` `0x00` to the communication server and then sends/receives segments (int24 length + data). Maybe you also will find out that the first 2 or 3 segments send/received are protobuf segments. After this handshake segments, all the other things are encrypted and it is not TLS, but the Noise protocol (as explained on the official WhatsApp website), so you cannot use MITM for that. The only way to get understand the handshake then is to decompile the WhatsApp APK file (Android). You will find out that the 2 or 3 handshake segments are Noise protocol handshake segments in protobuf format, that send an `ephemeral`, `static` and `payload` field. If you reverse engineer and implement the `WA` `0x03` `0x00` header and the handshake correctly, the WhatsApp server will send you code that you can easily decrypt with your own Noise protocol implementation client. However, this is still not the XML you want. You have to convert the bytes to their XML counterpart using a FunXMPP table (also in the WhatsApp APK or in this repository). Then you have the XML and you can understand the server. All the needed information is given here (endpoints, WhatsApp header, protobuf handshake format, FunXMPP tables), so the only thing to do is implement it: send `WA` `0x03` `0x00` to the server, implement the Noise protocol key exchange with the protobuf format, encrypt/decrypt your messages with the received private keys during this exchange, implement a FunXMPP encoder/decoder, and after that, you have a working WhatsApp client.
- I don't like Python that much, so I started my own project here (https://github.com/ben221199/WAPI) and I'm looking to yowsup how it all works. I don't know the future of yowsup, but I'm trying to bring my WAPI to live. I already implemented my client in Java with the Noise protocol and I'm receiving a `ServerHello` after sending my `ClientHello`. However, after my `ClientFinish`, I don't receive thing anymore. I'm finding out what goes wrong, if the Noise protocol used by WhatsApp is different from the standard and some more things. The goal of my https://github.com/ben221199/WAPI repository is to documentate the whole protocol and all stuff around it, not necessary implement a client, but I think I will post Java code in the future too.

**Comment by Satanjianpan** [comment]
Hello, I would like to ask about your project, can yowsup still be used normally?

I always encounter many problems

**Reply by ben221199** [reply]
It can, but if you encounter errors, you should look to the issues of this repository in order to fix it. If you cannot find any solution, you can create an issue yourself.

**Comment by eferay93** [comment]
I wonder why it fails to send message to myself while I can send to others. Is there any information about this?
